### DISCUSSION TITLE
Scoping issue - problems to do a cross-ref of a variable with same name in two different contexts

### DISCUSSION THREAD

**Discussion by dpturibio** [textX/textX#336] [Q&A]
Hello guys, I would appreciate your help in order to resolve this cross-referencing issue. 
In the grammar bellow, I need to recognize var1, var2 and var3 as belonging to its context(ZventoA and ZventoB) without conflicts:

``` 
from textx.metamodel import metamodel_from_str 
from textx.exceptions import TextXError
from textx.model import get_parent_of_type

grammar = """
 F:
     events *= Event; 

 Event: 
     'event' name=ID '(' parameters+=Variable[','] ')' ':'
     argumentos+=Repetition[',']
     'end'
 ;
 Repetition:
     ID '=' paramName=[Variable];

 Variable: name=ID;
 """
try:
    meta = metamodel_from_str(grammar, autokwd=True)
    print("Meta-Model OK")
except TextXError as e:
    print("Error in meta-model:")
    print(e)
    
code ='''
event ZeventA (var1,var2,var3):
    x1=var1,
    x2=var2,
    x3=var3
end

event ZeventoB (var1,var2,var3):
    x1=var1,
    x2=var2,
    x3=var3
end
'''
try:
    model = meta.model_from_str(code)
    print("Model ok")
except TextXError as e:
    print("Error in model: ", e)

```

But I have the following output when running the code above:
```
Meta-Model OK
Error in model:  None:3:8: error: name var1 is not unique.
```

When I change the name of var1 of event ZeventB to, for example ZeventB (var1aux,var2,var3), then the problem is transferred to var2

```
Meta-Model OK
Error in model:  None:4:8: error: name var2 is not unique.
```

How could I link var1, var2 and var3 to it's specific events ZeventA and ZeventB?
Thanks in advance.

**Comment by goto40** [comment]
Hi. Thank you for question.

Can you try with the fqn scope provider? This should work. http://textx.github.io/textX/stable/scoping/

**Comment by goto40** [comment]
Note: The fqn scope provider identifies each element by its full hierarchical name (e.g. 'ZeventA.var1'), but also allows relative names, like 'var1'.

**Comment by dpturibio** [comment]
> Hi. Thank you for question.
> 
> Can you try with the fqn scope provider? This should work. http://textx.github.io/textX/stable/scoping/

Thanks for the answer.
Is there any more detailed documentation to understand better how fqn scope provider works? I am a little confused yet in how to use in this case.

**Comment by goto40** [comment]
Have a look at the unittest referenced in the docs: https://github.com/textX/textX/blob/master/tests/functional/test_scoping/test_full_qualified_name.py

The FQN scope provider works much like scope provider in programming languages like Java works (look in the local scope first).

Note: you need to 'activate' the dot in the name ('FQN' rule). In your case, only names in the local scope are required. Thus, you can ommit the dot.

You are right, more docs would be nice. More or less related is this doc request: https://github.com/textX/textX/issues/102

**Comment by goto40** [comment]
I will prepare your example tonight or tomorrow...

**Comment by goto40** [comment]
Here is  the adapted example.
I added 
```
    from textx.scoping.providers import FQN
    meta.register_scope_providers({'*.*': FQN()})
```
and some asserts to demonstrate the functionality.

I hope my explanations above help to understand. Do not hesitate to ask more questions! Thank you for the interest.

```
from textx.metamodel import metamodel_from_str 
from textx.exceptions import TextXError
from textx.model import get_parent_of_type

grammar = """
 F:
     events *= Event; 

 Event: 
     'event' name=ID '(' parameters+=Variable[','] ')' ':'
     argumentos+=Repetition[',']
     'end'
 ;
 Repetition:
     ID '=' paramName=[Variable];

 Variable: name=ID;
 """
try:
    meta = metamodel_from_str(grammar, autokwd=True)
    from textx.scoping.providers import FQN
    meta.register_scope_providers({'*.*': FQN()})
    print("Meta-Model OK")
except TextXError as e:
    print("Error in meta-model:")
    print(e)
    
code ='''
event ZeventA (var1,var2,var3):
    x1=var1,
    x2=var2,
    x3=var3
end

event ZeventoB (var1,var2,var3):
    x1=var1,
    x2=var2,
    x3=var3
end
'''
try:
    model = meta.model_from_str(code)
    ZeventA=model.events[0]
    ZeventoB=model.events[1]
    assert ZeventA.name == 'ZeventA'
    assert ZeventoB.name == 'ZeventoB'
    assert ZeventA.argumentos[0].paramName == ZeventA.parameters[0]
    assert ZeventoB.argumentos[0].paramName == ZeventoB.parameters[0]
    assert ZeventoB.argumentos[0].paramName != ZeventA.parameters[0]
    print("Model ok")
except TextXError as e:
    print("Error in model: ", e)
```

**Comment by goto40** [comment]
Maybe you want to give you Repetitions a name (to identify the objects):
```
Repetition:
     name=ID '=' paramName=[Variable];
```

**Comment by dpturibio** [comment]
@goto40, thanks a lot for your help so far. Let me ask you one more thing. I have the following grammar:
```
grammar = """
 F:
     definitions *= Event1
     (start = Start )?
 ; 

 Event1: 
     'event' name=ID '(' parameters+=Variable[','] ')' ':'
     argumentos+=Repetition[',']
     'end'
 ;
 
 Start: 
     'start' name=ID 'link' event=[Event1] '('   
     argumentos+=Repetition[','] ')'
     'end'
 ;
 
 Repetition:
    name=ID '=' paramName=[Variable];

 Variable: name=ID;
 """
```

used the metamodel structure:

```
try:
    meta = metamodel_from_str(grammar)
    from textx.scoping.providers import FQN
    meta.register_scope_providers({'*.*': FQN()})
    print("Meta-Model OK")
except TextXError as e:
    print("Error in meta-model:", e)

```
and my model code as follows:

```
code ='''
event eventA (varA,varB,varC):
    x1=varA,
    x2=varB,
    x3=varC
end

start eventoB link eventA (
    x1=varA,
    x2=varB,
    x3=varC )
end
'''
try:
    model = meta.model_from_str(code)
    print("Model ok")
except TextXError as e:
    print("Error in model: ", e)

```

What should I do to make varA from "start" structure be linked with varA in "eventA" definition?
Because this code throws this exception: 
`Error in model:  None:9:8: error: Unknown object "varA" of class "Variable"`
What am I doing wrong(or not doing) in the code?

Thanks in advance.

**Comment by goto40** [comment]
You can use the RelativeName scope provider for such cases (references in Start). The difficulty in your case is that you have the Start and the Event1 both using Repetitions. I will analyze your example. I fear we need a custom scope provider...

Note:TEP-001 will probably make your example easy to specify...

**Comment by goto40** [comment]
Added example, see #201 
Is is quite simple to select a scope provider according to the underlying data structure.

@igordejanovic: Although TEP-001 will probably ease such things...

**Comment by goto40** [comment]
Merged #201: Now you have a new demo as unittest [`test_combined_scope_provider`](https://github.com/textX/textX/blob/master/tests/functional/test_scoping/test_issue170_combined_scope_provider.py) (your example), which defines a combined scope provider (`MyScope`):

Depending of the underlying data structure (`if textx_isinstance(obj.parent, self.meta["Start"]):`) one of two scope providers is selected.
```
class MyScope(object):
    """
    Demo of a combined scope provider:
    Dependent of the underlying data structure an appropriate
    provider is selected...
    """
    def __init__(self, meta):
        self.meta = meta
        self.pStart = RelativeName('parent.event.parameters')
        self.pEvent1 = RelativeName('parent.parameters')
        # also possible: self.pEvent1 = FQN()

    def __call__(self, obj, attr, obj_ref):
        if textx_isinstance(obj.parent, self.meta["Start"]):
            return self.pStart(obj, attr, obj_ref)
        else:
            return self.pEvent1(obj, attr, obj_ref)
```
This scope provider is allocated to `Repetition.paramName`: `'Repetition.paramName': MyScope(meta)`.

```
    meta = metamodel_from_str(grammar)
    meta.register_scope_providers({
        '*.*': FQN(),
        'Repetition.paramName': MyScope(meta)
    })
```

**Comment by goto40** [comment]
Note: In the new example we use the `RelativeName` scope provider to specify the "way through the model" to find the list of named objects to be used to resolve the reference. `RelativeName` also handles postponed references correctly. 

If you have a logic to decide what to do in your custom scope provider involving other references, then you need to handle unresolved references as well. In the given example, `obj.parent` (involved in the custom logic) has no reference involved, therefore we do not have to check for unresolved references...

**Comment by dpturibio** [comment]
Thank you very much for the example. It did the trick.
Another question: if there is more than 1 error in my code, just the first is thrown. Is there any way to obtain the subsequent ones?

**Comment by goto40** [comment]
At the moment there is no way. But you can make a feature request... It is a good idea!

**Comment by dpturibio** [comment]
Issue resolved. Thanks for @goto40

**Comment by goto40** [comment]
We do not close question issues to make them more visible...
